### PR TITLE
ShaderIR: Implement Shader Subfunctions. [WIP]

### PR DESCRIPTION
**Pull request by FernandoS27**
This PR implements `CAL` and `RET` instructions within our shader engine. I had made this a time ago and never finished it. @ReinUsesLisp and @bunnei  have promised to take care of it. So I'm leaving it as WIP for review.

This should fix some issues in HW: AoC & DQ11

### CODE REVIEW

File: src/video_core/renderer_vulkan/vk_shader_decompiler.cpp [hunk 1]
```diff
@@ -406,10 +406,38 @@ class SPIRVDecompiler final : public Sirit::Module {
         binding = DeclareStorageTexels(binding);
         binding = DeclareImages(binding);
 
+        const auto& subfunctions = ir.GetSubFunctions();
```
Comment (lioncash) [inline src/video_core/renderer_vulkan/vk_shader_decompiler.cpp:409]: Perhaps move this block into its own named function and call it here

File: src/video_core/shader/control_flow.cpp [hunk 2]
```diff
@@ -26,17 +27,29 @@ using Tegra::Shader::OpCode;
...
+enum class JumpLabel : u32 {
+    SSYClass = 0,
+    PBKClass = 1,
+};
+
+struct JumpItem {
+    JumpLabel type;
+    u32 address;
+
+    bool operator==(const JumpItem& other) const {
```
Comment (lioncash) [inline src/video_core/shader/control_flow.cpp:39]:
```suggestion
    auto operator<=>(const JumpItem& other) const = default;
```

File: src/video_core/shader/control_flow.cpp [hunk 3]
```diff
@@ -65,20 +78,36 @@ struct BlockInfo {
     }
 };
 
+struct ProgramControl {
+    std::unordered_set<u32> found_functions{};
+    std::list<u32> pending_functions{};
+
+    void RegisterFunction(u32 address) {
+        if (found_functions.count(address) != 0) {
```
Comment (lioncash) [inline src/video_core/shader/control_flow.cpp:86]:
```suggestion
        if (found_functions.contains(address)) {
```

File: src/video_core/shader/decode.cpp [hunk 4]
```diff
@@ -64,9 +64,52 @@ std::optional<u32> TryDeduceSamplerSize(const SamplerEntry& sampler_to_deduce,
...
+    }
+
+    void operator()(const ExprPredicate& expr) {
+        const auto pred = static_cast<Tegra::Shader::Pred>(expr.predicate);
+        if (pred != Pred::UnusedIndex && pred != Pred::NeverExecute) {
+            ir.used_predicates.insert(pred);
+        }
+    }
+
+    void operator()(const ExprCondCode& expr) {}
```
Comment (lioncash) [inline src/video_core/shader/decode.cpp:92]: Consider marking these as `[[maybe_unused]]` or not providing a parameter name to prevent unused parameter warnings.

File: src/video_core/shader/decode.cpp [hunk 5]
```diff
@@ -125,77 +178,113 @@ class ASTDecoder {
...
+        if (settings.depth != shader_info.settings.depth) {
+            LOG_WARNING(
+                HW_GPU,
+                "Decompiling to this setting \"{}\" failed, downgrading to this setting \"{}\"",
+                CompileDepthAsString(settings.depth),
+                CompileDepthAsString(shader_info.settings.depth));
+        }
+    });
+    const auto gen_function =
+        ([this](ShaderFunction& shader_info, u32 id) -> std::shared_ptr<ShaderFunctionIR> {
```
Comment (lioncash) [inline src/video_core/shader/decode.cpp:253]: The initial set of parentheses here isn't necessary.

File: src/video_core/shader/shader_ir.h [hunk 6]
```diff
@@ -186,7 +218,16 @@ class ShaderIR final {
         return num_custom_variables;
     }
 
+    std::shared_ptr<ShaderFunctionIR> GetMainFunction() const {
```
Comment (lioncash) [inline src/video_core/shader/shader_ir.h:221]:
```suggestion
    std::shared_ptr<ShaderFunctionIR>& GetMainFunction() const {
```

This should also be nodiscard, ditto for the const variant.

File: src/video_core/renderer_opengl/gl_arb_decompiler.cpp [hunk 7]
```diff
@@ -807,13 +810,33 @@ ARBDecompiler::ARBDecompiler(const Device& device_, const ShaderIR& ir_, const R
...
+    InitializeVariables();
+    AddLine("main:");
+    if (context_func->IsDecompiled()) {
         DecompileAST();
     } else {
         DecompileBranchMode();
+        AddLine("RET;");
+    }
+
+    const auto& subfunctions = ir.GetSubFunctions();
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_arb_decompiler.cpp:828]: Maybe place this block in its owned named function and call it here, like `DecompileSubfunctions();` or whatever is more appropriate.

File: src/video_core/renderer_opengl/gl_arb_decompiler.cpp [hunk 8]
```diff
@@ -807,13 +810,33 @@ ARBDecompiler::ARBDecompiler(const Device& device_, const ShaderIR& ir_, const R
     : device{device_}, ir{ir_}, registry{registry_}, stage{stage_} {
     DefineGlobalMemory();
 
+    context_func = ir.GetMainFunction();
+    ast_var_base = 0;
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_arb_decompiler.cpp:814]: `ast_var_base` is already set to zero inside the class body, so this line isn't necessary

File: src/video_core/renderer_opengl/gl_arb_decompiler.cpp [hunk 9]
```diff
@@ -807,13 +810,33 @@ ARBDecompiler::ARBDecompiler(const Device& device_, const ShaderIR& ir_, const R
     : device{device_}, ir{ir_}, registry{registry_}, stage{stage_} {
     DefineGlobalMemory();
 
+    context_func = ir.GetMainFunction();
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_arb_decompiler.cpp:813]: This can be moved into the constructor initializer list.

**Comment by lat9nq** [comment]
Causes a regression in FE:TH -- kills the radeonsi driver during battle.

[210115_feth120.tar.gz](https://github.com/yuzu-emu/yuzu/files/5976644/210115_feth120.tar.gz)
[yuzu_log.txt](https://github.com/yuzu-emu/yuzu/files/5976645/yuzu_log.txt)

**Comment by Xcedf** [comment]
Luigi's Mansion 3 issue https://github.com/yuzu-emu/yuzu/issues/5961 is here

**Comment by v1993** [comment]
Just wondering - what's the intended future of this PR? Is it planned to be superseded by shader decompiler rewrite and closed?

**Comment by FernandoS27** [comment]
> Just wondering - what's the intended future of this PR? Is it planned to be superseded by shader decompiler rewrite and closed?

To be honest this code was not meant to be released originally. I made it around HW:AOC release time as a proof of concept of fixes. I made a few test builds for internal use back then and it didn't resolve the issues of AoC so it was stashed. I was mostly retired due and I didn't continue it further into a proper solution.

It was pointed out by @BSoDGamingYT that this code fixed games like Dragon Quest and some UE4 games. So I was asked by @bunnei to PR it while we do a proper solution (aka project Hades).

TL;DR: This is not a proper solution or ever meant to be. Will be fully replaced by Project Hades aka New Shader Recompiler.
